### PR TITLE
Picard refactoring

### LINKED ISSUE #9038: Steady and eigen executioners do not have Picard iteration
### Description of the enhancement or error report
Should we add a simple outer Picard loop?

### Rationale for the enhancement or information for reproducing the error
This is required by some of the steady-state multiphysics simulations. Also if this is in place, codes in Rattlesnake for doing nonlinear diffusion acceleration can be cleaned up.

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)
New capability so no impact on the current capabilities.

### PR DESCRIPTION
**Pull request by YaqiWang**
This PR moves the Picard iteration into `FEProblem` in order to support Picard iteration for all executioners #9038. It clears things up significantly imo. I will give more explanations later. Just fire this up in our test box to see the impact.

This closes #9038 and #12255.

### CODE REVIEW

**Comment by permcody** [comment]
I like your ambition. I see several cleanups in the Transient loop too.

**Comment by YaqiWang** [comment]
This changes contains several steps (or commits):
1. detangle Picard iteration from others. This includes moving timestepper pre-, post-solve outside Picard iteration, moving time step setup outside, moving norm change evaluation for steady-state detection evaluation outside, moving the convergence check outside, etc. This step has to be done carefully. No tests should fail due to this change.
2. add new solver in `FEProblemBase`. This new solver will basically mimic the detangled Picard iteration in `Transient`.
3. switch on the new solver. `XFEM` and `Relaxation` need special care.

This is very important to us because more and more multiphysics tasks fall upon us requiring a robust Picard iteration implementation. I guess this will make time stepper refactoring, unionized executioner, etc. easier later. This will fail Rattlesnake badly and I or @vincentlaboure need more time to work on a patch. This will be a part of efforts for upgrading Rattlesnake to the new eigenvalue executioner.

**Comment by YaqiWang** [comment]
Thanks @permcody. If our past gold file are really *gold*, this refactoring should not need regolding. Some tests for guarding the old designs will need to be changed. So on the one hand, lots of regression tests makes this change really scary but on the other hand, they also makes this change right, or maybe less scary.

**Comment by permcody** [comment]
The Transient refactoring is tricky. I have a branch that I could share with you that was really close. I think it was only failing like one or two tests in the entire suite but I just never finished it. I have a feeling that you are going to run into recover problems with your loop refactoring. That was the trickiest part. Let me push up my ancient branch so you can see what I did, it might help to get the last few details right.

**Comment by permcody** [comment]
Oh you are right, recover passed! Good job!

**Comment by YaqiWang** [comment]
Wow, looks like the impact is much less than I imagined. Will take a look on the discussion. Last night I stopped investigating on why switching to the new solver for `Steady` caused one more time step in two tests (in my commit of "regolding a test"). Could be related. The level-set failure is due to my first commit, causing reentering `Transient::execute` to exhibit different behavior. The failure in statistical model is due to changes in `Steady` executioner. Those can be fixed easily. Impact on Bison is minor.

**Comment by permcody** [comment]
I just rebased my branch:

Here are the test results:
```
restart/restart_subapp_not_master.two_step_solve_master_restart .............................. FAILED (ERRMSG)
postprocessors/num_adaptivity_cycles.num_adaptivity_cycles_toggle_adaptivity ................ FAILED (CSVDIFF)
postprocessors/num_adaptivity_cycles.num_adaptivity_cycles_toggle_adaptivity_wait ........... FAILED (CSVDIFF)
multiapps/stateful_material_sub_cycling.test_stateful_subcycle .............................. FAILED (CSVDIFF)
multiapps/sub_cycling_failure.test_failure .................................................. FAILED (EXODIFF)
multiapps/picard_failure.test_csv ........................................................... FAILED (CSVDIFF)
multiapps/picard_failure.test ............................................................... FAILED (EXODIFF)
multiapps/picard_failure.test_2subapps ...................................................... FAILED (EXODIFF)
multiapps/catch_up.test ..................................................................... FAILED (EXODIFF)
executioners/adapt_and_modify.test .......................................................... FAILED (EXODIFF)
--------------------------------------------------------------------------------------------------------------
Ran 1876 tests in 97.9 seconds.
1866 passed, 77 skipped, 0 pending, 10 FAILED
```

... and in modules:
```
tensor_mechanics/test:crystal_plasticity.test_cutback ....................................... FAILED (EXODIFF)
combined/test:fdp_geometric_coupling.test ................................................... FAILED (EXODIFF)
xfem/test:second_order_elements.diffusion_quad9_levelsetcut ................................. FAILED (EXODIFF)
xfem/test:second_order_elements.diffusion_2d_quad9 .......................................... FAILED (EXODIFF)
xfem/test:second_order_elements.diffusion_2d_tri6 ........................................... FAILED (EXODIFF)
xfem/test:second_order_elements.diffusion_2d_quad8 .......................................... FAILED (EXODIFF)
xfem/test:diffusion_xfem.levelsetcut2d ...................................................... FAILED (EXODIFF)
xfem/test:diffusion_xfem.levelsetcut2d_aux .................................................. FAILED (EXODIFF)
xfem/test:diffusion_xfem.levelsetcut3d ...................................................... FAILED (EXODIFF)
xfem/test:solid_mechanics_basic.crack_propagation_sm ........................................ FAILED (EXODIFF)
xfem/test:single_var_constraint_2d.propagating_2field_1constraint ........................... FAILED (EXODIFF)
xfem/test:single_var_constraint_2d.propagating_1field ....................................... FAILED (EXODIFF)
xfem/test:single_var_constraint_2d.propagating_2field_2constraint ........................... FAILED (EXODIFF)
xfem/test:moving_interface.moving_diffusion ................................................. FAILED (EXODIFF)
xfem/test:moving_interface.moving_interface ................................................. FAILED (EXODIFF)
xfem/test:solid_mechanics_basic.crack_propagation_single_point .............................. FAILED (EXODIFF)
xfem/test:moving_interface.phase_transition ................................................. FAILED (EXODIFF)
xfem/test:moving_interface.moving_bimaterials ............................................... FAILED (EXODIFF)
misc/test:dynamic_loading/dynamic_load_multiapp.dynamic_load ................................ FAILED (EXODIFF)
phase_field/test:phase_field_kernels.ACInterfaceStress ...................................... FAILED (EXODIFF)
--------------------------------------------------------------------------------------------------------------
Ran 2597 tests in 304.0 seconds.
2577 passed, 204 skipped, 0 pending, 20 FAILED
```

**Comment by YaqiWang** [comment]
Since now timestep setup stuff is outside Picard iteration, fixing #12255 can be done easily.

**Comment by YaqiWang** [comment]
XFEM is interacting `Transient` in a bad manner. It should possibly just use Picard iteration for time-step resolving. In my change, I had to leave a hole for XFEM to work. Ideally, there should be none related XFEM in `Transient`.

**Comment by permcody** [comment]
I completely agree that XFEM should not be there. We will have to allow that hack for now until we can work with the developers who put it there.

File: framework/src/executioners/Transient.C [hunk 1]
```diff
@@ -371,25 +316,21 @@ Transient::execute()
   // in the future eventually.
   if (!_app.isRecovering())
     _problem.advanceState();
+  else
+    incrementStepOrReject();
 
   // Start time loop...
-  while (true)
+  while (keepGoing())
```
Comment (YaqiWang) [inline framework/src/executioners/Transient.C]: @permcody You will like this. This is the clean loop you wanted long ago ;-). This is in my first commit. It only fails one test in level-set module. I guess our past creeping changes in `Transient` made this easier.

**Comment by YaqiWang** [comment]
I think this is ready for review since moose framework and modules are working fine. I will take a look tomorrow on few apps to see what moose tests is missing. Tag @permcody. I might take a stab on removing the extra mesh adaptation at the end. One thing I do want to note is that we assume timestep_begin and timestep_end are picardstep_begin and picardstep_end. This makes a postprocessor getting number of Picard iterations at timestep_end tricky. I am not sure if we want to keep this assumption. I guess overall it is not bad.

**Comment by YaqiWang** [comment]
@permcody something is not quite right in few phase-field user objects. They are checking the solver status and switch calculations based on that. Those user objects could be running on `timestep_begin` where solver has not been called at the current time step. A better logic could exist. I assume this code is relevant with @dschwen. Checking solver status in moose user objects could result unexpected behavior because the objects can be executed at the place where checking solver status does not make sense.

**Comment by permcody** [comment]
I agree that checking converged on `timestep_begin` is a little odd, but the error here is related to compilation, not runtime logic.

**Comment by YaqiWang** [comment]
@permcody is it possible for you to give me the right to look into the failure in 'Private App tests'? I was missing a small logic with XFEM in the new Transient, that are captured by one Bison test. Bighorn test seems can be just regolded (tiny exodiffs). I will have to work on a temporary patch for Rattlesnake/MAMMOTH. Maggpie needs to be patched similarly as phase-field module, but I do not have access.

File: framework/src/problems/FEProblemBase.C [hunk 2]
```diff
@@ -4179,6 +4184,348 @@ FEProblemBase::solve()
...
+  picard_it = 0;
+  while (picard_it < solver_params._picard_max_its)
+  {
+    if (solver_params._has_picard_its)
+    {
+      _console << COLOR_MAGENTA << "Beginning Picard Iteration " << picard_it << COLOR_DEFAULT
+               << '\n';
+
+      if (picard_it ==
+          0) // First Picard iteration - need to save off the initial nonlinear residual
```
Comment (permcody) [inline framework/src/problems/FEProblemBase.C]: This wrapping is odd, perhaps move the comment up above the if statement?

File: framework/src/problems/FEProblemBase.C [hunk 3]
```diff
@@ -4179,6 +4184,348 @@ FEProblemBase::solve()
...
+          break;
+        }
+        if (picard_it + 1 == solver_params._picard_max_its)
+        {
+          _picard_status = MOOSE_PICARD_DIVERGED_MAX_ITS;
+          converged = false;
+          break;
+        }
+      }
+    }
```
Comment (permcody) [inline framework/src/problems/FEProblemBase.C]: This looks pretty good, however we could probably still continue to factor out more of this code into methods just so these loops and branch fit on a developer's screen when they are looking at this code. You don't necessarily need to do that in this PR. There's a lot of refactoring going on already.

File: framework/src/timesteppers/AB2PredictorCorrector.C [hunk 4]
```diff
@@ -85,8 +85,7 @@ AB2PredictorCorrector::step()
   NonlinearSystemBase & nl = _fe_problem.getNonlinearSystemBase();
   AuxiliarySystem & aux = _fe_problem.getAuxiliarySystem();
 
-  _fe_problem.solve();
-  _converged = _fe_problem.converged();
+  _converged = _fe_problem.newSolve();
```
Comment (permcody) [inline framework/src/timesteppers/AB2PredictorCorrector.C]: This is a little odd. We have newSolve() that returns a converged status but also a converged() call in NonlinearSystemBase()? Are we sure that's where we want the converge() method or should we just leave it at the FEProblem(Base) level?

File: modules/phase_field/src/userobjects/DiscreteNucleationFromFile.C [hunk 5]
```diff
@@ -103,7 +103,7 @@ DiscreteNucleationFromFile::initialize()
   _changes_made = {0, 0};
 
   // expire entries from the local nucleus list (if the current time step converged)
-  if (_fe_problem.converged())
+  if (_fe_problem.checkConvergence() > 0)
```
Comment (permcody) [inline modules/phase_field/src/userobjects/DiscreteNucleationFromFile.C]: Wait, also a checkConvergence() call too?

**Comment by permcody** [comment]
Attention @idaholab/moose-team -  This PR contains a very large refactoring effort in the Executioners. We are moving much of the solve logic into problem so that it can be shared among Executioners. One of the things I really like is that it cleans up the Transient loop, which means we'll be able to fix the extra adaptivity step at the end of the simulation a little more naturally, plus the loop just looks nicer. 

Disadvantages: This will certainly break existing Executioners. There's not a very clean way to maintain backwards compatibility, but fixing them up shouldn't be too much work. 

If you have any objections to this PR - let's hear about them soon before this gets merged.

File: framework/include/problems/FEProblemBase.h [hunk 6]
```diff
@@ -127,6 +127,18 @@ enum MooseLinearConvergenceReason
...
+  MOOSE_PICARD_ITERATING = 0,
+  MOOSE_PICARD_CONVERGED_NONLINEAR = 1,
+  MOOSE_PICARD_CONVERGED_ABS = 2,
+  MOOSE_PICARD_CONVERGED_RELATIVE = 3,
+  MOOSE_PICARD_CONVERGED_CUSTOM = 4,
+  MOOSE_PICARD_DIVERGED_MAX_ITS = -1,
+  MOOSE_PICARD_DIVERGED_NONLINEAR = -2,
+  MOOSE_PICARD_DIVERGED_FAILED_MULTIAPP = -3
+};
+
```
Comment (fdkong) [inline framework/include/problems/FEProblemBase.h]: Moving it to MooseType might be better.

File: framework/include/problems/FEProblemBase.h [hunk 7]
```diff
@@ -127,6 +127,18 @@ enum MooseLinearConvergenceReason
   MOOSE_DIVERGED_PCSETUP_FAILED = -11
 };
 
+enum MoosePicardConvergenceReason
```
Comment (permcody) [inline framework/include/problems/FEProblemBase.h]: You should consider the new strong type `enum class` here.

File: framework/include/problems/FEProblemBase.h [hunk 8]
```diff
@@ -127,6 +127,18 @@ enum MooseLinearConvergenceReason
...
+{
+  MOOSE_PICARD_ITERATING = 0,
+  MOOSE_PICARD_CONVERGED_NONLINEAR = 1,
+  MOOSE_PICARD_CONVERGED_ABS = 2,
+  MOOSE_PICARD_CONVERGED_RELATIVE = 3,
+  MOOSE_PICARD_CONVERGED_CUSTOM = 4,
+  MOOSE_PICARD_DIVERGED_MAX_ITS = -1,
+  MOOSE_PICARD_DIVERGED_NONLINEAR = -2,
+  MOOSE_PICARD_DIVERGED_FAILED_MULTIAPP = -3
```
Comment (fdkong) [inline framework/include/problems/FEProblemBase.h]: Moving it to MooseType might be better?

File: framework/src/executioners/Executioner.C [hunk 9]
```diff
@@ -203,7 +267,7 @@ Executioner::getTimeStepperName()
 bool
 Executioner::lastSolveConverged()
 {
-  return _fe_problem.converged();
+  return true;
```
Comment (fdkong) [inline framework/src/executioners/Executioner.C]: We alway converge?

**Comment by permcody** [comment]
Note: @YaqiWang and I discussed. The new API is not going to be "newSolve()", this is a temporary placeholder. It will be "solve()".

File: framework/src/multiapps/MultiApp.C [hunk 10]
```diff
@@ -144,6 +145,15 @@ validParams<MultiApp>()
...
 
+  params.addRangeCheckedParam<Real>("relaxation_factor",
+                                    1.0,
+                                    "relaxation_factor>0 & relaxation_factor<2",
+                                    "Fraction of newly computed value to keep."
+                                    "Set between 0 and 2.");
+  params.addParam<std::vector<std::string>>("relaxed_variables",
+                                            std::vector<std::string>(),
+                                            "List of variables to relax during Picard Iteration");
+
```
Comment (rwcarlsen) [inline framework/src/multiapps/MultiApp.C]: Why is this here?

File: framework/src/multiapps/MultiApp.C [hunk 11]
```diff
@@ -661,6 +670,18 @@ MultiApp::createApp(unsigned int i, Real start_time)
...
+  auto & solver_params = fe_problem_base.solverParams();
+  solver_params._picard_self_relaxation_factor = getParam<Real>("relaxation_factor");
+  solver_params._picard_self_relaxed_variables =
+      getParam<std::vector<std::string>>("relaxed_variables");
+  if (solver_params._picard_self_relaxation_factor != 1.0)
+  {
+    // Store a copy of the previous solution here
+    fe_problem_base.getNonlinearSystemBase().addVector("self_relax_previous", false, PARALLEL);
+    fe_problem_base.previousTimeAsSubapp() = fe_problem_base.time() - 1;
+  }
```
Comment (rwcarlsen) [inline framework/src/multiapps/MultiApp.C]: This seems weird - it shouldn't go here me thinks.

**Review by rwcarlsen — CHANGES_REQUESTED** [review]
Is there really no better way to address this than creating another `newSolve`?  - there has to be a less inelegant way.  If we keep the newSolve approach, there definitely needs to be a detailed plan of how things are migrated/deprecated/updated.

File: framework/src/problems/FEProblemBase.C (hunk 2 above)
Comment (YaqiWang) [reply framework/src/problems/FEProblemBase.C]: wow, this is done by the formatter. I will move the inline comments down inside the if branch.

File: framework/src/timesteppers/AB2PredictorCorrector.C (hunk 4 above)
Comment (YaqiWang) [reply framework/src/timesteppers/AB2PredictorCorrector.C]: This is the complicated part. What I have done in this PR is that when `FEProblemBase::solve` is done, it returns true/false to indicate if the solve successful or not. It also set the status of the solver `_picard_status` like how the solve is converged or diverged, which can be inquired by `FEProblemBase::checkConvergence`. In the future, users may want to check more status of the solve, like convergence history, etc, so keeping `FEProblemBase::checkConvergence` makes sense. Returning true/false makes things handy although we can definitely use `FEProblemBase::checkConvergence` to inquire the convergence.

I guess leaving `NonlinearSystemBase::converged` which indicates if nonlinear solver is converged is fine, plus phase-field module and few apps are directly checking that.

File: modules/phase_field/src/userobjects/DiscreteNucleationFromFile.C (hunk 5 above)
Comment (YaqiWang) [reply modules/phase_field/src/userobjects/DiscreteNucleationFromFile.C]: Yes, this is the issue I raised: checking solver status is not safe inside moose object because solver status may not be valid at the point of checking.

File: framework/src/multiapps/MultiApp.C (hunk 10 above)
Comment (YaqiWang) [reply framework/src/multiapps/MultiApp.C]: It is complicated. Let us say we have a master + subapp + subapp. When we do Picard relaxation, we can
1. do relaxation on master;
2. do relaxation on subapp as a master of another subapp;
3. do relaxation on subapp as subapp of its master.
This is used to control Case 3.

File: framework/include/problems/FEProblemBase.h (hunk 8 above)
Comment (YaqiWang) [reply framework/include/problems/FEProblemBase.h]: We have `MooseNonlinearConvergenceReason` and `MooseLinearConvergenceReason` sitting right above this. Maybe move all of them?

File: framework/src/executioners/Executioner.C (hunk 9 above)
Comment (YaqiWang) [reply framework/src/executioners/Executioner.C]: I possibly should make this pure virtual.

File: framework/src/multiapps/MultiApp.C (hunk 11 above)
Comment (YaqiWang) [reply framework/src/multiapps/MultiApp.C]: This is the trick used in the old implementation. It is used to check if we are re-entering `newSolve` in the same time step, which can happen with Picard iteration on master with subapps.

**Comment by YaqiWang** [comment]
We will not have `newSolove` in this PR. It is a holder for `solve`.

File: modules/phase_field/src/userobjects/DiscreteNucleationFromFile.C (hunk 5 above)
Comment (permcody) [reply modules/phase_field/src/userobjects/DiscreteNucleationFromFile.C]: It's not that it's invalid, it's just that the result may not be expected. Perhaps the converged flag should be reset every time we advance timesteps? If we did that and somebody had a user object checking the flag before the solve, they would get an "unsolved or not converged" flag back and then they'd have to think a little bit about what they were trying to do. I wouldn't over engineer this. It might just be worth talking to the developers or looking at the code to understand the intent.

**Comment by rwcarlsen** [comment]
@YaqiWang - I know - what I was asking was are there no better ways to make the transition than via the temporary `newSolve`?  If not, then what is the plan for transitioning everything and getting rid of newSolve?

**Comment by YaqiWang** [comment]
There is another option to keep `FEProblemBase` still doing the nonlinear solve and the picard iteration in `Executioner` (the base class). I did not do it because the current way give developers less chance of messing up the executioners. This actually makes patching Rattlesnake difficult. Rattlesnake has an executioner altering Picard iterations quite intensively. I will have to derive a new FEProblem for doing that same calculation with the same `Transient` in MOOSE.

File: framework/include/problems/FEProblemBase.h (hunk 8 above)
Comment (permcody) [reply framework/include/problems/FEProblemBase.h]: I guess it depends on how wide of scope we need. If we really don't need these status flags outside of the problem/executioner, they are fine here. Putting everything in the types file for the sake of having them all in one place may not be the best approach. I would say that this is fine if the other statuses are also currently sitting here.

File: framework/src/executioners/Executioner.C (hunk 9 above)
Comment (permcody) [reply framework/src/executioners/Executioner.C]: Probably - aren't they overridden in our Executioners anyway?

**Comment by YaqiWang** [comment]
We will do immediate transition. I have another commit sitting in my branch to rename `newSolve` to `solve` and remove the old solve in `FEProblemBase`.

File: framework/src/executioners/Executioner.C (hunk 9 above)
Comment (YaqiWang) [reply framework/src/executioners/Executioner.C]: Yes. Now `FEProblemBase` does not own this flag. When it finish solving, it hands the flag to the caller. In `Transient`, the time stepper is holding the flag and in `Steady`, the executioner is holding. We could use `FEProblemBase::checkConvergence` here and then all the MOOSE executioners will no longer implement this. We can still make it virtual to give chance the derived executioners overriding it.

File: framework/include/problems/FEProblemBase.h (hunk 8 above)
Comment (YaqiWang) [reply framework/include/problems/FEProblemBase.h]: Will do.

File: framework/include/problems/FEProblemBase.h (hunk 7 above)
Comment (YaqiWang) [reply framework/include/problems/FEProblemBase.h]: I can use this for all three enums here in FEProblemBase.h.

File: modules/phase_field/src/userobjects/DiscreteNucleationFromFile.C (hunk 5 above)
Comment (YaqiWang) [reply modules/phase_field/src/userobjects/DiscreteNucleationFromFile.C]: Looks like only @dschwen is doing this. I will create a new enum `MOOSE_PICARD_UNSOLVED` and set solver status in `FEProblemBase::timestepSetup`.

File: framework/src/executioners/Executioner.C (hunk 9 above)
Comment (YaqiWang) [reply framework/src/executioners/Executioner.C]: Looks like time steppers have legislated reason to decide if the solve is converged or not. So I will just make this function pure virtual and force derived executioners to implement it.

**Comment by YaqiWang** [comment]
Comments are addressed. I will need to add few more tests for steady-state Picard iterations.

**Comment by YaqiWang** [comment]
Need to fix the compiling error in modules. Working on another test to let `Eigenvalue` executioner drive the picard iteration. I will push the test later and possibly need @fdkong to take a look.

**Comment by fdkong** [comment]
> @fdkong The test in the last commit behaves strangely. It is a test derived from existing `tests/problems/eigen_problem/ne_coupled.i` by split the system into two files. One contains the eigenvalue problem without `T` variable, another solve `T` only. The two files are linked with multiapp and transfers. Picard iteration is turned on in the master eigenvalue problem.
> 
> Now I am getting an error the second time calling the eigenvalue problem solve:
> 
> ```
> [0]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
> [0]PETSC ERROR: Shift-and-invert requires a target 'which' (see EPSSetWhichEigenpairs), for instance -st_type sinvert -eps_target 0 -eps_target_magnitude
> [0]PETSC ERROR: See http://www.mcs.anl.gov/petsc/documentation/faq.html for trouble shooting.
> [0]PETSC ERROR: Petsc Release Version 3.10.1, unknown 
> [0]PETSC ERROR: ../../../moose_test-opt on a arch-darwin-c-opt named inl651110 by wangy2 Thu Jan 24 00:28:16 2019
> [0]PETSC ERROR: Configure options --download-hypre=1 --with-debugging=no --with-shared-libraries=1 --with-64-bit-indices=true --download-fblaslapack=1 --download-metis=1 --download-parmetis=1 --download-superlu_dist=1 --with-mpi=1 PETSC_ARCH=arch-darwin-c-opt PETSC_DIR=/Users/wangy2/projects/petsc
> [0]PETSC ERROR: #1 EPSSetUp() line 272 in /Users/wangy2/projects/slepc/src/eps/interface/epssetup.c
> [0]PETSC ERROR: #2 EPSSolve() line 135 in /Users/wangy2/projects/slepc/src/eps/interface/epssolve.c
> Segmentation fault: 11
>

The `eigen solver`  change can be made in a different PR. Ricard refactoring itself is already huge.

**Comment by YaqiWang** [comment]
Using T system as mater, I am getting:
```
[0]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
[0]PETSC ERROR: Argument out of range
[0]PETSC ERROR: Too many EPS monitors set
[0]PETSC ERROR: See http://www.mcs.anl.gov/petsc/documentation/faq.html for trouble shooting.
[0]PETSC ERROR: Petsc Release Version 3.10.2, unknown 
[0]PETSC ERROR: ../../../moose_test-opt on a arch-moose named inl601947 by wangy2 Thu Jan 24 12:08:10 2019
[0]PETSC ERROR: Configure options --download-hypre=1 --with-ssl=0 --with-debugging=no --with-pic=1 --with-shared-libraries=1 --with-cc=mpicc --with-cxx=mpicxx --with-fc=mpif90 --download-fblaslapack=1 --download-metis=1 --download-ptscotch=1 --download-parmetis=1 --download-superlu_dist=1 --download-mumps=1 --download-scalapack=1 -CC=mpicc -CXX=mpicxx -FC=mpif90 -F77=mpif77 -F90=mpif90 -CFLAGS="-fPIC -fopenmp" -CXXFLAGS="-fPIC -fopenmp" -FFLAGS="-fPIC -fopenmp" -FCFLAGS="-fPIC -fopenmp" -F90FLAGS="-fPIC -fopenmp" -F77FLAGS="-fPIC -fopenmp"
[0]PETSC ERROR: #1 EPSMonitorSet() line 83 in /Users/wangy2/projects/slepc/src/eps/interface/epsmon.c
[0]PETSC ERROR: #2 EPSConvMonitorSetFromOptions() line 87 in /Users/wangy2/projects/slepc/src/eps/interface/epsopts.c
[0]PETSC ERROR: #3 EPSSetFromOptions() line 246 in /Users/wangy2/projects/slepc/src/eps/interface/epsopts.c
```
the third time calling eigen solver.

**Comment by YaqiWang** [comment]
@permcody all comments are addressed. The tests should pass now. If no further comments, this can be merged and we can start fixing apps.

**Comment by YaqiWang** [comment]
Fixed a small typo in SQA requirement.

**Comment by YaqiWang** [comment]
Summarize apps that need to be patched:

Magpie - failed due to calling `FEProblem::converged`, can be fixed as phase-field module;
Prariedog - same as Magpie;
Okapi - using a derived executioner that needs to be updated, @lindsayad has the patch;
I cannot see the test results of private apps, but they probably just need some trivial fixes as above.

Rattlesnake patch is not easy. It has a derived executioner from `Transient` doing some Picard manipulations. I am working on a quick temporary fix. `Non unity build` failure is due the the incompatible usage of interfaces in the executioner.

**Comment by YaqiWang** [comment]
Any further comments?

**Review by friedmud — CHANGES_REQUESTED** [review]
Ok - here's the deal... I like the idea of disentangling the Transient solver from the Picard solver.  I think that's good.

BUT: I don't think the Picard solver should get put into FEProblem.  An FEProblem should just be _one_ solve of one finite-element problem.  It shouldn't have this kind of looping in it.  Picard/Transient/Adaptivity loops should be _external_ to the FEProblem and just ask it to solve.

So: What needs to be done here is that all of the Picard stuff needs to get put into a `PicardSolve` object... that can be used by both the Transient and Steady Executioners.

**Comment by friedmud** [comment]
Well sue me.  I was busy and on travel.  How about before you go changing
something huge like this you run it past me first - especially if you're
worried that you might do a lot of work and then have me not agree.

I still stand firm with my statements here.

One of the big reasons why I don't think this is the right way to do things
is because of convergence testing.  I see that you've drilled down to the
Nonlinear solver in several places to see if it converged... when that's
not the right thing to do.  FEProblem::converged() is where apps can put in
their own convergence checking of the nonlinear problem.

By putting Picard inside FEProblem we've now mixed together two different
ideas.  What does "converged()" mean for the FEProblem now?

Plus: we need to be pulling things OUT of FEProblem... not putting them in.

Pulling out the PicardSolve object will achieve all goals: it will keep it
from bloating FEProblem, it will make it usable in multiple places
(Transient and Steady) and it will keep a clear separation between what's
being solved in an FEProblem and what's going on in the Picard iterations
(providing two places for convergence checks for instance).

**Comment by YaqiWang** [comment]
If you were not watching MOOSE I would be very surprised cause even I am doing. Plus I'v mentioned this idea in @permcody 's PR during a conversation. Anyway this is not the point. I am not sure if you noticed the conversation in this PR, in which I mentioned that phase-field module and few other apps have objects directly inquiring the convergence of FEProblem. It is not the right thing to do but I have to put a bandage on for the old behavior. Another reason makes me think this is the right thing is that users can derive their own FEProblem and a general PicardSolve object may not be sufficient to their FEProblem and they may be forced to create dedicated Executioners for their own FEProblem. Putting this into one place in FEProblem is thus desirable. I agree we should reduce the size of FEProblem, thinking carefully what should be included in FEProblem and what the purpose it is, but this is something I think belonging to FEProblem.

**Comment by YaqiWang** [comment]
Really need to get this done so that we can move forward to upgrade Rattlesnake with the new eigen executioner (also per @rcm59 request). We need to reach a consensus quickly, tagging @idaholab/moose-team. I will make changes accordingly either way. I believe I have made my point clearly. If you have any further questions, please let me know.

**Comment by permcody** [comment]
The comments on this ticket aren't constructive. I believe the Actions we need to take lie somewhere in the middle. Anyone who wants a voice in the direction of the framework development really should be subscribed to the repository and responsive to tickets being opened, PRs being opened, etc. I do somewhat agree with that statement. However, in this particular instance, the design and code was completed as part of the very first PR testing so comments in this instance wouldn't have been possible before the code was designed. 

Let's work together and get this done.

The fact of the matter is that disruptive changes in the framework take time. A little back and forth is going to occur no matter how much code and how much testing occurs. Thanks for your understanding.

**Comment by YaqiWang** [comment]
Currently `TimeStepper::step` is calling `FEProblem::solve`, which is wrapped inside Picard iteration in `Transient`. The change I made is to put Picard iteration inside `FEProblem::solve` and still let `TimeStepper::step` call `FEProblem::solve`. I think letting `step` take care of Picard iteration as a whole is a logically good move. If we want the Picard iteration in somewhere other than `FEProblem::solve`, we will need to let `TimeStepper` know where the code is and call it properly, which possibly will result some unnecessary dependencies among classes.

**Comment by friedmud** [comment]
> If you were not watching MOOSE I would be very surprised cause even I am doing

What part of "I was traveling" did you not understand?  The day you put this in (without talking to any of us) I was literally flying to Chicago - where I was working with people all week.

At any rate: I'm not beholden to you so I don't need to explain myself.

BTW: I can put my opinion in about code at any point in time... even after it's been merged.  The collective here can feel free to ignore my opinion... but it's just as valid at any point in the development.

I would like to get into the same room to understand your needs better here and come up with a design that everyone can agree on.

**Comment by YaqiWang** [comment]
Superseded by #12797.
